File: unit-testing-workshop/test/com/ihsinformatics/unittesting/HelloUnitTest.java
/**
 * 
 */
package com.ihsinformatics.unittesting;

import static org.junit.Assert.fail;

import java.math.BigInteger;
import java.util.stream.Stream;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;

/**
 * @author owais.hussain@ihsinformatics.com
 *
 */
public class HelloUnitTest {

	/**
	 * @throws java.lang.Exception
	 */
	@BeforeClass
	public static void setUpBeforeClass() throws Exception {
		System.out.println("setUpBeforeClass() is executed before first test");
	}

	/**
	 * @throws java.lang.Exception
	 */
	@AfterClass
	public static void tearDownAfterClass() throws Exception {
		System.out.println("tearDownAfterClass() is executed after all tests");
	}

	/**
	 * @throws java.lang.Exception
	 */
	@Before
	public void runBeforeEachTest() throws Exception {
		System.out.println("runBeforeEachTest() is executed before each test");
	}

	/**
	 * @throws java.lang.Exception
	 */
	@After
	public void runAfterEachTest() throws Exception {
		System.out.println("runAfterEachTest() is executed after each test");
	}

	@Test
	public void boundToFail() {
		fail("boundToFail() will just fail the test case");
	}

	@Test(expected = NullPointerException.class)
	public void shouldThrowNullPointerException() {
		Double.parseDouble(null);
	}

	@Test(timeout = 2000)
	public void shouldNotTakeMoreThanTwoSeconds() {
		for (int i = 0; i < 10000000; i++) {
			BigInteger bigInteger = new BigInteger(String.valueOf(i));
			bigInteger.multiply(bigInteger);
		}
	}

	@Test
	public void shouldConvertDouble() {
		try {
			Stream.of("0", "1.2", "2", "3.33D", "3.14F", "-1", "-99.99", "1234567890.987654321")
					.forEach(num -> Double.parseDouble(num));
		} catch (Exception e) {
			fail("Should convert all the numbers.");
		}
	}


	@Test
	@Ignore
	public void ignoreMe() {
		if (1 > 0) {
			fail("I don't existed");
		}
	}
}

File: unit-testing-workshop/src/python/hello_unit_test.py
import time
import unittest


class HelloUnitTest(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        print("setUpClass() is executed before first test")

    @classmethod
    def tearDownClass(cls):
        print("tearDownClass() is executed after all tests")

    def setUp(self):
        print("setUp() is executed before each test")

    def tearDown(self):
        print("tearDown() is executed after each test")

    def test_bound_to_fail(self):
        self.fail("test_bound_to_fail() will just fail the test case")

    def test_should_throw_type_error(self):
        with self.assertRaises(TypeError):
            float(None)

    def test_should_not_take_more_than_twenty_seconds(self):
        start = time.monotonic()
        for i in range(10_000_000):
            i * i
        elapsed = time.monotonic() - start
        self.assertLess(elapsed, 20.0, msg=f"Took {elapsed:.2f}s, expected < 20s")

    def test_should_convert_double(self):
        numbers = ["0", "1.2", "2", "3.33", "3.14", "-1", "-99.99", "1234567890.987654321"]
        try:
            for num in numbers:
                float(num)
        except Exception:
            self.fail("Should convert all the numbers.")

    @unittest.skip("Demonstrates @Ignore equivalent")
    def test_ignore_me(self):
        if 1 > 0:
            self.fail("I don't existed")


if __name__ == "__main__":
    unittest.main()
